pkg/pipeline/stream/destination_acker.go

```diff
@@ -34,114 +32,117 @@ type DestinationAckerNode struct {
+		})
 	}()
 
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-n.stop:
-		// destination actually stopped without ever receiving a message, we can
-		// just return here
-		return nil
-	case <-n.start:
-		// received first message for ack, destination is open now, we can
-		// safely start listening to acks
-		n.logger.Trace(ctx).Msg("start running acker node")
+	trigger, cleanup, err := n.base.Trigger(ctx, n.logger, errChan)
+	if err != nil {
+		close(errChan) // need to close errChan to not block deferred function
+		return err
 	}
 
+	// start worker that will fetch acks from the connector and forward them to
+	// internal messages
+	go n.worker(connectorCtx, signalChan, errChan)
+
+	defer cleanup()
 	for {
-		pos, err := n.Destination.Ack(ctx)
-		if pos == nil {
-			// empty position is returned only if an actual error happened
-			if cerrors.Is(err, plugin.ErrStreamNotOpen) {
-				// this means the plugin stopped, gracefully shut down
-				n.logger.Debug(ctx).Msg("ack stream closed")
-				return nil
-			}
+		msg, err := trigger()
+		if err != nil || msg == nil {
 			return err
 		}
 
-		msg, ok := n.cache.LoadAndDelete(pos)
-		if !ok {
-			n.logger.Error(ctx).
-				Str(log.RecordPositionField, pos.String()).
-				Msg("received unexpected ack (could be an internal bug or a badly written connector), ignoring the ack and continuing, please report the issue to the Conduit team")
-			continue
-		}
-
-		// TODO make sure acks are called in the right order or this will block
-		//  forever. Right now we rely on connectors sending acks back in the
-		//  correct order and this should generally be true, but a badly written
-		//  connector could provoke a deadlock, we could prevent that.
-		err = n.handleAck(msg, err)
-		if err != nil {
-			return err
+		n.m.Lock()
+		n.queue.PushBack(msg)
+		n.m.Unlock()
+		select {
+		case signalChan <- struct{}{}:
+			// triggered the start of listening to acks in worker goroutine
+		default:
+			// worker goroutine is already busy, it will pick up the message
+			// because it is already stored in the queue
 		}
 	}
 }
 
-// teardown will nack all messages still in the cache and return an error in
-// case there were still unprocessed messages in the cache.
-func (n *DestinationAckerNode) teardown(reason error) error {
-	var nacked int
-	var err error
-	n.cache.Range(func(pos record.Position, msg *Message) bool {
-		err = multierror.Append(err, msg.Nack(reason))
-		nacked++
-		return true
-	})
-	if err != nil {
-		return cerrors.Errorf("nacked %d messages when stopping destination acker node, some nacks failed: %w", nacked, err)
+func (n *DestinationAckerNode) worker(
+	ctx context.Context,
+	signalChan <-chan struct{},
+	errChan chan<- error,
+) {
+	handleError := func(msg *Message, err error) {
+		// push message back to the front of the queue and return error
+		n.m.Lock()
+		n.queue.PushFront(msg)
```

Comment:
Exactly, it needs to be pushed into the queue to be nacked in teardown. The order definitely plays a role, because acks/nacks need to be processed in order (remember the semaphore?), that's why we need to push it to the front, otherwise we could have a deadlock.

Reply:
I was thinking that we can stop with the first failed message, since the error will stop the pipeline anyway, but since the queue may have other messages in it, all of them need to be taken care of in proper order.